netboot/netconf.go

```diff
@@ -74,6 +78,91 @@ func GetNetConfFromPacketv6(d *dhcpv6.DHCPv6Message) (*NetConf, error) {
 	return &netconf, nil
 }
 
+// GetNetConfFromPacketv4 extracts network configuration information from a DHCPv4
+// Reply packet and returns a populated NetConf structure
+func GetNetConfFromPacketv4(d *dhcpv4.DHCPv4) (*NetConf, error) {
+	// extract the address from the DHCPv4 address
+	ipAddr := d.YourIPAddr()
+	if ipAddr.Equal(net.IPv4zero) {
+		return nil, errors.New("ip address is null (0.0.0.0)")
+	}
+	netconf := NetConf{}
+
+	// get the subnet mask from OptionSubnetMask. If the netmask is not defined
+	// in the packet, an error is returned
+	netmaskOption := d.GetOneOption(dhcpv4.OptionSubnetMask)
+	if netmaskOption == nil {
+		return nil, errors.New("no netmask option in response packet")
+	}
+	netmask := netmaskOption.(*dhcpv4.OptSubnetMask).SubnetMask
+	if binary.LittleEndian.Uint32(netmask) == 0 {
+		return nil, errors.New("netmask extracted from OptSubnetMask options is null")
+	}
+
+	// netconf struct requires a valid lifetime to be specified. ValidLifetime is a dhcpv6
+	// concept, the closest mapping in dhcpv4 world is "IP Address Lease Time". If the lease
+	// time option is nil, we set it to 0
+	leaseTimeOption := d.GetOneOption(dhcpv4.OptionIPAddressLeaseTime)
+	leaseTime := uint32(0)
+	if leaseTimeOption != nil {
+		leaseTime = leaseTimeOption.(*dhcpv4.OptIPAddressLeaseTime).LeaseTime
+	}
+
+	if int(leaseTime) < 0 {
+		return nil, fmt.Errorf("lease time overflow, Original lease time: %d", leaseTime)
+	}
+
+	netconf.Addresses = append(netconf.Addresses, AddrConf{
+		IPNet: net.IPNet{
+			IP:   ipAddr,
+			Mask: netmask,
+		},
+		PreferredLifetime: 0,
+		ValidLifetime:     int(leaseTime),
+	})
+	if bytes.Equal(netmask, []byte{0, 0, 0, 0}) {
+		return nil, errors.New("netmask in response packet is null")
+	}
+
+	// get DNS configuration
+	dnsServersOption := d.GetOneOption(dhcpv4.OptionDomainNameServer)
+	if dnsServersOption == nil {
+		return nil, errors.New("no dns server option in response packet")
+	}
+	dnsServers := dnsServersOption.(*dhcpv4.OptDomainNameServer).NameServers
+	if len(dnsServers) == 0 {
+		return nil, errors.New("no dns servers options in response packet")
```

Comment:
nit: this message may be confusing with the one above, maybe "name servers list is empty" or similar?